zephyr: Implement block device protocol and enable file systems

MaureenHelm opened the pull request:
Introduces two new classes that implement the block device protocol, and enables the fatfs and littlefs file systems in the zephyr port.

The `zephyr.Disk` class uses the zephyr disk access api and supports only the standard block device interface (without the offset parameter). It can be used with SDHC, RAM, or FLASH disk access drivers and the fatfs file system. See the git log for an example demonstrating how to use it with the SDHC driver on the mimxrt1050_evk board.

The `zephyr.Flash` class uses the zephyr flash map api and supports both the standard and extended block device interfaces. It can be used with FLASH drivers, and the fatfs or littlefs file systems. See the git log for an example demonstrating how to use it on the reel_board, rv32m1_vega_ri5cy, or frdm_k64f board.

This is a simpler alternative to #5292

cc: @yanyao12138

MaureenHelm committed:
zephyr: Implement block device protocol via zephyr disk access api.

Introduces a new zephyr.DiskAccess class that uses the zephyr disk
access api to implement the uos.AbstractBlockDev protocol. This can be
used with any type of zephyr disk access driver, which currently
includes SDHC, RAM, and FLASH implementations. The SDHC driver is
enabled on the mimxrt1050_evk board.

Only the standard block device protocol (without the offset parameter)
can be supported with the zephyr disk access api, therefore this class
cannot be used with file systems like littlefs which require the
extended interface. In the future it may be possible to implement the
extended interface in a new class using the zephyr flash api.

MaureenHelm committed:
zephyr: Enable virtual file system and uos module.

Enables the virtual file system and uos module in the zephyr port.
No concrete file system implementations are enabled yet.

MaureenHelm committed:
zephyr: Enable fatfs.

Enables the fatfs filesystem in the zephyr port.

Example usage with an SD card on the mimxrt1050_evk board:

import zephyr, os
bdev = zephyr.DiskAccess('SDHC')
os.VfsFat.mkfs(bdev)
os.mount(bdev, '/sd')
with open('/sd/hello.txt','w') as f:
    f.write('Hello world')
print(open('/sd/hello.txt').read())

MaureenHelm committed:
zephyr: Implement block device protocol via zephyr flash map api.

Introduces a new zephyr.FlashArea class that uses the zephyr flash map
api to implement the uos.AbstractBlockDev protocol. The flash driver is
enabled on the frdm_k64f board, reel_board, and rv32m1_vega_ri5cy board.

The standard and extended block device protocols are both supported,
therefore this class can be used with file systems like littlefs which
require the extended interface.

MaureenHelm committed:
zephyr: Enable littlefs.

Enables the littlefs (v1 and v2) filesystems in the zephyr port.

Example usage with the internal flash on the reel_board or the
rv32m1_vega_ri5cy board:

import os
from zephyr import FlashArea
bdev = FlashArea(FlashArea.STORAGE, 4096)
os.VfsLfs2.mkfs(bdev)
os.mount(bdev, '/flash')
with open('/flash/hello.txt','w') as f:
    f.write('Hello world')
print(open('/flash/hello.txt').read())

Things get a little trickier with the frdm_k64f due to the micropython
application spilling into the default flash storage partition defined
for this board. The zephyr build system doesn't enforce the flash
partitioning when mcuboot is not enabled (which it is not for
micropython). For now we can demonstrate that the littlefs filesystem
works on frdm_k64f by constructing the FlashArea block device on the
mcuboot scratch partition instead of the storage partition. Do this by
replacing the FlashArea.STORAGE constant above with the value 4.

MaureenHelm commented:
@dpgeorge any feedback?

MaureenHelm commented:
@dpgeorge ping

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+
+STATIC void zephyr_disk_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    zephyr_disk_obj_t *self = self_in;
+    mp_printf(print, "%s", self->pdrv);
```
Perhaps print `Disk("%s")` instead of just the path?

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_WRITE);
+    ret = disk_access_read(self->pdrv, bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / self->block_size);
+    return mp_obj_new_bool(ret == 0);
```
This function can return an int, so just returning `mp_obj_new_small_int(ret);` would be simpler and allow more information to pass to the upper layers in case of error.

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_READ);
+    ret = disk_access_write(self->pdrv, bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / self->block_size);
+    return mp_obj_new_bool(ret == 0);
```
Same as above, can return int.

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+
+        case MP_BLOCKDEV_IOCTL_SYNC:
+            if (disk_access_ioctl(self->pdrv, DISK_IOCTL_CTRL_SYNC, &buf)) {
+                mp_raise_ValueError("unable to sync");
```
Probably makes sense to return an error integer rather than raise here.  Maybe just return the int returned from this call?

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+            return MP_OBJ_NEW_SMALL_INT(buf);
+
+        case MP_BLOCKDEV_IOCTL_BLOCK_SIZE:
+            if (disk_access_ioctl(self->pdrv, DISK_IOCTL_GET_SECTOR_SIZE, &buf)) {
```
Does it make sense to just return the cached value in `self->block_size`?  That shouldn't change over time, and means there can be no error here.

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+            if (disk_access_ioctl(self->pdrv, DISK_IOCTL_GET_SECTOR_COUNT, &buf)) {
+                mp_raise_ValueError("unable to get block count");
+            }
+            return MP_OBJ_NEW_SMALL_INT(buf);
```
Does it make sense to call `disk_access_ioctl(GET_SECTOR_COUNT)` in the constructor (make_new) and cache the value and return it here?  Like with `self->block_size`.

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+
+STATIC const mp_obj_type_t zephyr_disk_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_Disk,
```
Could this class/type be called `DiskAccess` instead to better map to the Zephyr name?  I have no strong feelings either way, but it seems that `Disk` is very generic and maybe `DiskAccess` signals to the user that it's a more special thing with a specific meaning.

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+
+        case MP_BLOCKDEV_IOCTL_BLOCK_ERASE:
+            if (flash_area_erase(self->area, block_num * self->block_size, self->block_size)) {
+                mp_raise_ValueError("unable to erase block");
```
Return an error code instead of raising?

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
+
+STATIC const mp_obj_type_t zephyr_flash_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_Flash,
```
As above with `Disk`, maybe call this class `FlashArea` to map to the Zephyr name?

dpgeorge commented on ports/zephyr/mpconfigport.h:
```diff
@@ -80,6 +81,16 @@
+#define MICROPY_FATFS_LFN_CODE_PAGE    437 /* 1=SFN/ANSI 437=LFN/U.S.(OEM) */
+#define MICROPY_FATFS_USE_LABEL        (1)
+#define MICROPY_FATFS_RPATH            (2)
+#define MICROPY_FATFS_MULTI_PARTITION  (1)
```
I don't think MULTI_PARTITION is needed.  (It's enabled on stm32 to support mounting SD cards with a partition table. But that's at the C level, the Python level can't make use of multi-partition in the FAT driver.)

dpgeorge commented:
Thanks a lot @MaureenHelm for continuing on from #5292, and sorry for the slow response.

What you've done here is really great, I think the approach of using block devices will be much more familiar to users that have experience with other MicroPython ports.  I'd be happy to merge this, once the minor details are sorted out (see above review comments).

Also, I think it'd be good to move the new Disk/Flash code from `modzephyr.c` to a separate file (but still stay in the `zephyr` module), because otherwise `modzephyr.c` will eventually grow too large.  Maybe something like `zephyr_storage.c` following how other ports split their module implementation up.

A few other general points/questions:
- How does Zephyr support Littlefs on SDHC?  (does it?)  Because that's not possible with the implementation here.
- In principal we can still support this, with a small Python shim between the `Disk` class and the littlefs driver to deal with the offset within the block (it may need a block cache though).  Or could just use littlefs with read/write block size equal to the block size of the `Disk` object.
- Note that it's still possible to add VfsPosix support (option 2 in #5292) on top of what is done here, if that becomes desirable in the future.

MaureenHelm commented:
@dpgeorge Thanks for the review!

Rebased to latest master (to pick up #5477) and incorporated all requested changes.
 
> Also, I think it'd be good to move the new Disk/Flash code from `modzephyr.c` to a separate file (but still stay in the `zephyr` module), because otherwise `modzephyr.c` will eventually grow too large. Maybe something like `zephyr_storage.c` following how other ports split their module implementation up.

Done
 
> * How does Zephyr support Littlefs on SDHC?  (does it?)  Because that's not possible with the implementation here.

It's not possible in zephyr either. The Littlefs integration inside zephyr uses the flash area interface, but SDHC drivers implement the disk access interface. There is a bridge available between the two interfaces, but it only goes one way (it implements disk access on top of flash, which allows mounting FatFS on flash)

> * Note that it's still possible to add VfsPosix support (option 2 in #5292) on top of what is done here, if that becomes desirable in the future.

Thanks. I still don't quite understand what this means, but I need to do some homework.

dpgeorge commented on ports/zephyr/main.c:
```diff
@@ -132,16 +136,26 @@ void gc_collect(void) {
+#endif
 
 mp_import_stat_t mp_import_stat(const char *path) {
+#if MICROPY_VFS
```
code style: we try to indent preprocessor directives to the level of the code it protects, so here it would have one level of indentation (4 spaces)

dpgeorge commented on ports/zephyr/main.c:
```diff
@@ -132,16 +136,26 @@ void gc_collect(void) {
 }
 
 mp_obj_t mp_builtin_open(size_t n_args, const mp_obj_t *args, mp_map_t *kwargs) {
+#if MICROPY_VFS
```
as above, these 3 new directives would be indented one level

dpgeorge commented on ports/zephyr/moduos.c:
```diff
@@ -0,0 +1,76 @@
+
+STATIC const mp_rom_map_elem_t uos_module_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_uos) },
+#if MICROPY_VFS
```
indented

dpgeorge commented on ports/zephyr/moduos.c:
```diff
@@ -0,0 +1,76 @@
+    { MP_ROM_QSTR(MP_QSTR_mount), MP_ROM_PTR(&mp_vfs_mount_obj) },
+    { MP_ROM_QSTR(MP_QSTR_umount), MP_ROM_PTR(&mp_vfs_umount_obj) },
+#endif
+#if MICROPY_VFS_FAT
```
indented

dpgeorge commented on ports/zephyr/moduos.c:
```diff
@@ -0,0 +1,76 @@
+#if MICROPY_VFS_FAT
+    { MP_ROM_QSTR(MP_QSTR_VfsFat), MP_ROM_PTR(&mp_fat_vfs_type) },
+#endif
+#if MICROPY_VFS_LFS1
```
indented

dpgeorge commented on ports/zephyr/moduos.c:
```diff
@@ -0,0 +1,76 @@
+#if MICROPY_VFS_LFS1
+    { MP_ROM_QSTR(MP_QSTR_VfsLfs1), MP_ROM_PTR(&mp_type_vfs_lfs1) },
+#endif
+#if MICROPY_VFS_LFS2
```
indented

dpgeorge commented on ports/zephyr/modzephyr.c:
```diff
@@ -31,6 +32,7 @@
 #include <zephyr.h>
 #include <debug/stack.h>
 
+#include "zephyr_storage.h"
```
while the .c source is `zephyr_storage.c` the header file would still be `modzephyr.h` which would collect all exported items for all things related to the `zephyr` module (this helps to keep the number of .h files down, since .h's are usually much smaller than .c they don't need to be split up, and it helps also to logically collect the definitions in one place)

dpgeorge commented on ports/zephyr/prj_mimxrt1050_evk.conf:
```diff
@@ -0,0 +1,3 @@
+# Required for zephyr.Disk block devices
```
i think the comment here should read "zephyr.DiskAccess"

dpgeorge commented on ports/zephyr/zephyr_storage.c:
```diff
@@ -0,0 +1,268 @@
+    ret = disk_access_read(self->pdrv, bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / self->block_size);
+    return MP_OBJ_NEW_SMALL_INT(ret);
+}
+
```
we usually don't put an empty line between the function and the wrapper just below, to indicate that they have a very close relationship

dpgeorge commented on ports/zephyr/zephyr_storage.h:
```diff
@@ -0,0 +1,35 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
```
as mentioned above, this file would be better named `modzephyr.h`; and for consistency have the following guard (see eg modmachine.h):
```c
#ifndef MICROPY_INCLUDED_ZEPHYR_MODZEPHYR_H
#define MICROPY_INCLUDED_ZEPHYR_MODZEPHYR_H
...
#endif // MICROPY_INCLUDED_ZEPHYR_MODZEPHYR_H
```

dpgeorge reviewed (commented):
Thanks for updating the PR.  I've made some additional review comments which are very minor, all basically style/convention items (which are hard for contributors to know about because the code base is not 100% consistent with the use of them).  Usually I would fix such things myself during rebase and merge but just wanted to point them out here first.  Feel free to fix them, or otherwise I'm happy to do it myself during merge.

dpgeorge commented:
Also, commit comments should end with a period "." (and be 72 chars or less).

MaureenHelm commented:
> Thanks for updating the PR. I've made some additional review comments which are very minor, all basically style/convention items (which are hard for contributors to know about because the code base is not 100% consistent with the use of them). Usually I would fix such things myself during rebase and merge but just wanted to point them out here first. Feel free to fix them, or otherwise I'm happy to do it myself during merge.

Updated. I think I got them all. I didn't rebase this time so you can see what changed in the force-pushed link above.

Also took the opportunity to reread `CODECONVENTIONS.md`. :)

dpgeorge commented:
Thanks for updating, and for the very clean PR!  Rebased without changes and merged in cc19cf2549d4cc8fc24bc35d8ba2187b23a6ab74 through 7a5752a7489f6be1c7307455b33119888392a09d
